enable I2C support on imx8mp Cortex-A Core

JiafeiPan opened the pull request:
1. New Zephyr driver for II2C on imx8m series platforms.
2. Add i2c clock support for control driver mcux_ccm.
3. Add i2c device nodes in imx8mp dts.
4. Enable i2c on imx8mp_evk A53 board.

JiafeiPan committed:
drivers: i2c: add NXP i2c driver used on imx8m platforms

This native i2c driver works together with ii2c drive in hal_nxp.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
drivers: clock_control: mcux_ccm: add ii2c clock support

Add I2C clock support for imx8m platforms.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
dts: arm64: imx8mp: add i2c dts nodes

Add I2C dts nodes for imx8mp Cortex-A Core platform.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
boards: imx8mp_evk: add i2c support

Added i2c3 support on imx8mp_evk A53 board.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

pdgendt commented on drivers/i2c/i2c_nxp_ii2c.c:
```diff
@@ -0,0 +1,416 @@
+	return 0;
+}
+
+static const struct i2c_driver_api nxp_ii2c_driver_api = {
```
```suggestion
static DEVICE_API(i2c, nxp_ii2c_driver_api) = {
```

JiafeiPan replied:
Done, thanks.

JiafeiPan commented:
Rebased.

JiafeiPan commented:
rebased and refined the patch.

JiafeiPan commented:
rebased and fix compliance check error.

dbaluta commented on drivers/i2c/CMakeLists.txt:
```diff
@@ -69,6 +69,7 @@ else()
 endif()
 zephyr_library_sources_ifdef(CONFIG_I2C_NUMAKER		i2c_numaker.c)
+zephyr_library_sources_ifdef(CONFIG_I2C_NXP_II2C	i2c_nxp_ii2c.c)
 zephyr_library_sources_ifdef(CONFIG_I2C_SAM_TWI		i2c_sam_twi.c)
```
Is this a typo or the naming is actually II2c and not I2C?

JiafeiPan replied:
It is I2C, but we call this IP on imx8m series with the name ii2c, same as GPIO on imx8m series is igpio, we use the same driver name in hal_nxp.

mmahadevan108 commented on drivers/i2c/i2c_nxp_ii2c.c:
```diff
@@ -0,0 +1,416 @@
+/*
+ * Copyright 2024 NXP
```
2025

JiafeiPan replied:
Done, thanks.

mmahadevan108 commented on dts/bindings/i2c/nxp,ii2c.yaml:
```diff
@@ -0,0 +1,15 @@
+# Copyright 2024 NXP
```
Update year

JiafeiPan replied:
Done, thanks.

mmahadevan108 commented on dts/bindings/i2c/nxp,ii2c.yaml:
```diff
@@ -0,0 +1,15 @@
+
+compatible: "nxp,ii2c"
+
+include: [i2c-controller.yaml, pinctrl-device.yaml, "nxp,rdc-policy.yaml"]
```
Why is "nxp,rdc-policy.yaml"

mmahadevan108 commented on drivers/i2c/i2c_nxp_ii2c.c:
```diff
@@ -0,0 +1,416 @@
+
+	bitrate_cfg = i2c_map_dt_bitrate(config->bitrate);
+
+
```
You can delete some of these blank lines

JiafeiPan replied:
Done, thanks.

mmahadevan108 commented:
@JiafeiPan , HAL PR has been merged.

JiafeiPan commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -447,6 +459,31 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
 #endif /* CONFIG_SPI_MCUX_ECSPI */
+
+#if defined(CONFIG_I2C_NXP_II2C) && defined(CONFIG_SOC_SERIES_IMX8M)
+	case IMX_CCM_I2C1_CLK:
```
Done, thanks.

JiafeiPan commented:
> @JiafeiPan , HAL PR has been merged.

Addressed comments from @mmahadevan108, update west.yml to include hal driver update in hal_nxp.

JiafeiPan committed:
west.yml: update hal_nxp to include ii2c hal driver

Updated to include ii2c hal driver update commit:
be4d59d9c84e9519f3c05645f410b6dd63aeac4f

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

mmahadevan108 commented:
@teburd , can you help review.

mmahadevan108 commented:
@pdgendt @dbaluta , can you please re-review.

pdgendt commented on drivers/i2c/i2c_nxp_ii2c.c:
```diff
@@ -0,0 +1,414 @@
+
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+LOG_MODULE_REGISTER(i2c_nxp_ii2c);
```
No need to set the LOG_LEVEL?

pdgendt commented on drivers/i2c/i2c_nxp_ii2c.c:
```diff
@@ -0,0 +1,414 @@
+	ARG_UNUSED(handle);
+	ARG_UNUSED(base);
+
+	struct device *dev = userdata;
+	struct nxp_ii2c_data *data = dev->data;
```
Prefer to declare variables at the beginning of the scope.
